### PR TITLE
Small cleanup of CMake compiler flags + prep work for improving the CMake experience.

### PR DESCRIPTION
**Pull request by slurps-mad-rips**
:sparkles: Add `LIBUV_BUILD_BENCH` option (currently does nothing, but is intended to help with #1886 once implemented)

:sparkles: Add compiler flag checks for gcc-style warnings. This allows more compilers with a GCC compatible interface to be used.

:art: Remove hard-coded C89 (which is actually ISO C90) flag and set it via `CMAKE_C_STANDARD`. Additionally the standard is *required* and GNU Extensions are enabled.

:art: `CMAKE_C_VISIBILITY_PRESET` is now used to set symbols to hidden instead of a hardcoded compiler flag

:wrench: Some CMake policies are now enabled for just the project which will come in handy later.

:art: Most of uv_test_sources was primarily moved to be under `LIBUV_BUILD_TESTS` check. This will make working with the rest of the file a lot easier, and allow us to delay modifying how the tests are built.

### CODE REVIEW

**Comment by slurps-mad-rips** [comment]
I can't request reviewers, but figure a quick ping to @saghul @bnoordhuis for review can't hurt :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,22 +2,48 @@
 cmake_minimum_required(VERSION 3.4)
 project(libuv LANGUAGES C)
 
+cmake_policy(SET CMP0057 NEW)
```
Comment (saghul) [inline CMakeLists.txt]: Can you perhaps add a one-liner above each, explaining what they do? a link to the CMake docs is also a good option.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,22 +2,48 @@
...
+check_c_compiler_flag(-Wno-unused-parameter UV_LINT_NO_UNUSED_PARAMETER)
+
+set(lint-strict-prototypes $<$<BOOL:${UV_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
+set(lint-default $<$<BOOL:${UV_LINT_STRICT_DEFAULT}>:-Wall>)
+set(lint-extra $<$<BOOL:${UV_LINT_STRICT_EXTRA}>:-Wextra>)
+
+set(lint-no-unused-parameter $<$<BOOL:${UV_LINT_NO_UNUSED_PARAMETER}>:-Wno-unused-parameter>)
+
 if(MSVC)
   list(APPEND uv_cflags /W4)
```
Comment (saghul) [inline CMakeLists.txt]: can we also check for this one?

File: CMakeLists.txt [hunk 3]
```diff
@@ -2,22 +2,48 @@
...
+
+check_c_compiler_flag(-Wstrict-prototypes UV_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wextra UV_LINT_EXTRA)
+check_c_compiler_flag(-Wall UV_LINT_DEFAULT)
+
+check_c_compiler_flag(-Wno-unused-parameter UV_LINT_NO_UNUSED_PARAMETER)
+
+set(lint-strict-prototypes $<$<BOOL:${UV_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
+set(lint-default $<$<BOOL:${UV_LINT_STRICT_DEFAULT}>:-Wall>)
+set(lint-extra $<$<BOOL:${UV_LINT_STRICT_EXTRA}>:-Wextra>)
```
Comment (saghul) [inline CMakeLists.txt]: Is this The Right Way (TM) to do it now? I get the compiler agnostic argument, but it feels much more verbose: it's a 3 step process.

File: CMakeLists.txt [hunk 4]
```diff
@@ -2,22 +2,48 @@
...
+  "Build the benchmarks when building unit tests and we are the root project" ON
+  "LIBUV_BUILD_TESTS" OFF)
+
+check_c_compiler_flag(-Wstrict-prototypes UV_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wextra UV_LINT_EXTRA)
+check_c_compiler_flag(-Wall UV_LINT_DEFAULT)
+
+check_c_compiler_flag(-Wno-unused-parameter UV_LINT_NO_UNUSED_PARAMETER)
+
+set(lint-strict-prototypes $<$<BOOL:${UV_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
```
Comment (saghul) [inline CMakeLists.txt]: Where can I read some more on this syntax?

File: CMakeLists.txt (hunk 1 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: Sure thing! I'll push up some changes soon(ish)

File: CMakeLists.txt (hunk 2 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: We can, sure.

File: CMakeLists.txt (hunk 3 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: It's currently three steps, however I plan to put a lot of this boilerplate into a "UV.cmake" (or some other named file) so we can backport a few CMake bug fixes and make this sort of thing appear to be a one-line call.

File: CMakeLists.txt (hunk 4 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: These are CMake's [generator expressions][]. The best analogy I can give is that they are a very poor imitation of s-expressions. One thing not mentioned is that they do not support being written on multiple lines in place or in a call to set. The "best" way to write a multi-line generator expression is via `string(CONCAT)`. I have a few examples in an upcoming PR when we need to use an `$<OR>` expression, but I want to keep it readable.

Generator expressions are most useful in that they allow you to remove a majority of your configure time branching and push them into generation time. Configuration in CMake is single-threaded, while generation is multi-threaded.

[generator expressions]: https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html

File: CMakeLists.txt (hunk 3 above)
Comment (saghul) [reply CMakeLists.txt]: Me likey!

**Comment by bnoordhuis** [comment]
This PR has the same issue as your first one: there are too many changes and they're mashed into a single commit.

It's okay to make multiple changes in a PR but they should be separate commits, each with its own rationale in the commit log, bonus points if they can be cherry-picked independently.

(Apropos commit log rationale, I'm a stickler for documenting _what_ and _why_.)

**Comment by slurps-mad-rips** [comment]
Ahh ok. I have this bad habit of squashing changes before a PR. 😅

I'll get this cleaned up. Give me a day or so. I'm sort of (but also sort of not?) in the range of the bay area power outages you might have seen folks talking about on social media. 😬

**Comment by saghul** [comment]
No rush, good work! :-)

**Comment by slurps-mad-rips** [comment]
Broke up the commits as requested @bnoordhuis. I'd like to get one or two more commits in if that would be alright. I know the `uv_test_sources` bit is going to be most controversial, but it's a huge list that really does need to be moved so that the diffs are "less distracting" while I cleanup the rest of the file.

File: CMakeLists.txt (hunk 2 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: This is in the latest set of commits.

File: CMakeLists.txt (hunk 3 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: Still on the TODO, but it will be coming :)

File: CMakeLists.txt [hunk 5]
```diff
@@ -2,22 +2,54 @@
...
-  list(APPEND uv_cflags -Wno-unused-parameter)
-endif()
+# Compiler check
+string(CONCAT is-msvc $<OR:
+  $<C_COMPILER_ID:MSVC>,
+  $<STREQUAL:${CMAKE_C_SIMULATED_ID},MSVC>
+>)
+
+check_c_compiler_flag(/W4 UV_LINT_W4)
+check_c_compiler_flag(-Wall UV_LINT_WALL) # DO NOT use this under MSVC
```
Comment (saghul) [inline CMakeLists.txt]: why not?

File: CMakeLists.txt [hunk 6]
```diff
@@ -2,22 +2,54 @@
...
+# TODO: Place these into its own function
+check_c_compiler_flag(-Wno-unused-parameter UV_LINT_NO_UNUSED_PARAMETER)
+check_c_compiler_flag(-Wstrict-prototypes UV_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wextra UV_LINT_EXTRA)
+
+set(lint-no-unused-parameter $<$<BOOL:${UV_LINT_NO_UNUSED_PARAMETER}>:-Wno-unused-parameter) 
+set(lint-strict-prototypes $<$<BOOL:${UV_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
+set(lint-extra $<$<BOOL:${UV_LINT_EXTRA}>:-Wextra>)
+set(lint-w4 $<$<BOOL:${UV_LINT_W4}>:/W4>)
+# Unfortunately, this one is complicated because MSVC and clang-cl support -Wall
```
Comment (saghul) [inline CMakeLists.txt]: if it's supported, why not use it?

File: CMakeLists.txt [hunk 7]
```diff
@@ -2,22 +2,54 @@
...
+string(CONCAT lint-default $<
+  $<AND:$<BOOL:${UV_LINT_WALL}>,$<NOT:${is-msvc}>>:-Wall
+>)
+
+list(APPEND uv_cflags ${lint-strict-prototypes} ${lint-extra} ${lint-default} ${lint-w4})
+list(APPEND uv_cflags ${lint-no-unused-parameter})
```
Comment (saghul) [inline CMakeLists.txt]: can't these 2 lines be folded in to one?

**Comment by saghul** [comment]
> Broke up the commits as requested @bnoordhuis.

Thanks!

>  I'd like to get one or two more commits in if that would be alright.

As long as they are in the context of this PR, go for it!

>  I know the `uv_test_sources` bit is going to be most controversial, but it's a huge list that really does need to be moved so that the diffs are "less distracting" while I cleanup the rest of the file.

Personally I don't see a problem with that, makes sense!

File: CMakeLists.txt (hunk 7 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: I was keeping them separate in case more flags were desired, but yeah we definitely could!

File: CMakeLists.txt (hunk 6 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: It's equivalent to `-Weverything` in clang. It turns on ALL the warnings :(

Comment (saghul) [reply CMakeLists.txt]: Oh, got it. Maybe put it in a comment so people who also don't know will know :-)

File: CMakeLists.txt (hunk 7 above)
Comment (saghul) [reply CMakeLists.txt]: I think it's ok to have a single list with all the flags, we'll see how we cross that river when we get there I guess.

File: CMakeLists.txt [hunk 8]
```diff
@@ -24,13 +25,26 @@ cmake_dependent_option(LIBUV_BUILD_BENCH
...
   "LIBUV_BUILD_TESTS" OFF)
 
-if(MSVC)
-  list(APPEND uv_cflags /W4)
-elseif(CMAKE_C_COMPILER_ID MATCHES "AppleClang|Clang|GNU")
-  list(APPEND uv_cflags -Wall -Wextra -Wstrict-prototypes)
-  list(APPEND uv_cflags -Wno-unused-parameter)
+# Unfortunately, MSVC and clang-cl both support -Wall, but we don't actually
+# want that, as its equivalent to -Weverything :/
+if (MSVC OR CMAKE_C_SIMULATED_ID STREQUAL "MSVC")
```
Comment (bnoordhuis) [inline CMakeLists.txt]: Google turns up zero matches for `CMAKE_C_SIMULATED_ID`, that's not something you see everyday anymore. ^^

Can you elaborate little on what it does and where it comes from? (That CheckCCompilerFlag module?)

File: CMakeLists.txt [hunk 9]
```diff
@@ -24,13 +25,26 @@ cmake_dependent_option(LIBUV_BUILD_BENCH
...
+  check_c_compiler_flag(/W4 UV_LINT_DEFAULT)
+else()
+  check_c_compiler_flag(-Wall UV_LINT_DEFAULT)
 endif()
 
+# TODO: Place these into its own function
```
Comment (bnoordhuis) [inline CMakeLists.txt]: Because...?

File: CMakeLists.txt (hunk 4 above)
Comment (bnoordhuis) [reply CMakeLists.txt]: > they do not support being written on multiple lines in place or in a call to set

Does this work?
```cmake
set(lint-no-unused-parameter
     $<$<BOOL:${UV_LINT_NO_UNUSED_PARAMETER}>:-Wno-unused-parameter)
```
I'd prefer to keep lines <= 80 columns if at all possible.

File: CMakeLists.txt [hunk 10]
```diff
@@ -404,6 +247,164 @@ target_include_directories(uv_a PUBLIC include PRIVATE src)
 target_link_libraries(uv_a ${uv_libraries})
 
 if(LIBUV_BUILD_TESTS)
+  list(APPEND uv_test_sources
+      test/blackhole-server.c
```
Comment (bnoordhuis) [inline CMakeLists.txt]: Needs one more space of indent. :)

File: CMakeLists.txt (hunk 9 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: It's regarding an earlier conversion with @saghul further up. Basically there's a lot of boilerplate just to "check and add" a compiler flag, and this could be reduced down to a single function.

File: CMakeLists.txt (hunk 8 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: Ahh this appears to actually be a typo. The correct string is `CMAKE_C_SIMULATE_ID`. You can read more about it [here][1]

[1]: https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_SIMULATE_ID.html

**Comment by slurps-mad-rips** [comment]
@bnoordhuis gotcha. I can go through and re-edit the commits via a rebase to meet that requirement.

File: CMakeLists.txt (hunk 8 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: CMake, of course, makes it very difficult to find things sometimes 😅. The `CMAKE_C_SIMULATE_ID` is a more recent variable that is set. The reason we want to check for it is if someone is using clang-cl, which is the clang driver with an msvc style frontend. If someone uses an older version of cmake it won't work with clang-cl to begin with (for reasons I'd rather not go into 🙄), but this allows us to be forward compatible.

Comment (bnoordhuis) [reply CMakeLists.txt]: Ah, very nice. Thanks for explaining!

File: CMakeLists.txt (hunk 9 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: (I intend to have that function written before it gets merged, if that helps 😅)

**Comment by saghul** [comment]
> I can go through and re-edit the commits via a rebase to meet that requirement.

👍

**Comment by slurps-mad-rips** [comment]
Sorry for the long delay. Life has got me swamped. I'll be returning to this soon!

File: CMakeLists.txt (hunk 10 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: This is fixed as of my latest commit :)

**Comment by slurps-mad-rips** [comment]
@saghul @bnoordhuis aside from a few commit message rebasing, is there anything else I need to do? I'd like to get this merged and start work on a fresh PR is possible :)

File: CMakeLists.txt (hunk 7 above)
Comment (saghul) [reply CMakeLists.txt]: Can you please addresss this?

File: CMakeLists.txt [hunk 11]
```diff
@@ -395,25 +438,28 @@ if(LIBUV_BUILD_TESTS)
...
+if (UNIX)
   # Now for some gibbering horrors from beyond the stars...
-  foreach(x ${uv_libraries})
-    set(LIBS "${LIBS} -l${x}")
-  endforeach(x)
+  foreach(lib IN LISTS uv_libraries)
+    list(APPEND LIBS "-l${lib}")
+  endforeach()
+  string(REPLACE ";" " " LIBS "${LIBS}")
+  #Consider setting project version via project() call?
```
Comment (saghul) [inline CMakeLists.txt]: Nit: space after `#`

**Comment by bnoordhuis** [comment]
I think touch-ups + rebase + CI is all that remains. @saghul Agree/disagree?

File: CMakeLists.txt [hunk 12]
```diff
@@ -395,25 +438,28 @@ if(LIBUV_BUILD_TESTS)
...
-  set_target_properties(uv PROPERTIES VERSION ${UV_VERSION_MAJOR}.0.0
-				      SOVERSION ${UV_VERSION_MAJOR})
+  set_target_properties(uv PROPERTIES
+    VERSION ${UV_VERSION_MAJOR}.0.0
+    SOVERSION ${UV_VERSION_MAJOR})
   set(includedir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR})
   set(libdir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR})
   set(prefix ${CMAKE_INSTALL_PREFIX})
-  configure_file(libuv.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc @ONLY)
+  configure_file(libuv.pc.in libuv.pc @ONLY)
```
Comment (saghul) [inline CMakeLists.txt]: Won't this leave the new file in the source directory?

File: CMakeLists.txt [hunk 13]
```diff
@@ -2,22 +2,54 @@
...
+# Compiler check
+string(CONCAT is-msvc $<OR:
+  $<C_COMPILER_ID:MSVC>,
+  $<STREQUAL:${CMAKE_C_COMPILER_FRONTEND_VARIANT},MSVC>
+>)
+
+check_c_compiler_flag(/W4 UV_LINT_W4)
+check_c_compiler_flag(-Wall UV_LINT_WALL) # DO NOT use this under MSVC
+
+# TODO: Place these into its own function
```
Comment (saghul) [inline CMakeLists.txt]: Can this TODO be removed now?

File: CMakeLists.txt (hunk 12 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: No. The output for `configure_file` is always relative to the `CMAKE_CURRENT_BINARY_DIR`, unless it's an absolute path.

File: CMakeLists.txt (hunk 13 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: It'll be in my next commit!

File: CMakeLists.txt (hunk 11 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: I'll take care of it in my next commit.

File: CMakeLists.txt (hunk 7 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: Sure thing! Must have missed it 😅

File: CMakeLists.txt (hunk 12 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: More info can be found [here][]

> `<output>`
    Path to the output file or directory. A relative path is treated with respect to the value of CMAKE_CURRENT_BINARY_DIR. If the path names an existing directory the output file is placed in that directory with the same file name as the input file.

[here]: https://cmake.org/cmake/help/latest/command/configure_file.html

Comment (saghul) [reply CMakeLists.txt]: TIL, thank you!

File: CMakeLists.txt (hunk 13 above)
Comment (saghul) [reply CMakeLists.txt]: Next PR? Ok, then let's leave it.

Comment (slurps-mad-rips) [reply CMakeLists.txt]: (next commit because I have to rebase and redo the commit messages. The actual "moving into their own function" part will be in the next PR)

**Comment by saghul** [comment]
@slurps-mad-rips is this ready for merge or will you make any further changes here?

**Comment by slurps-mad-rips** [comment]
> @slurps-mad-rips is this ready for merge or will you make any further changes here?

I need to make a few more changes with commit messages and I think a comment or two. I'll be getting to it this weekend. Sadly my free time has been at an all time low, though this has changed in the past 24 hours :)

**Comment by saghul** [comment]
No worries, I had lost track myself so I wanted to touch base :-) Thanks a lot for your dedication here!

**Comment by slurps-mad-rips** [comment]
@bnoordhuis @saghul Let me know if any other changes are needed for comments or commit messages.

Also if we should squash everything into one commit, I'm ok with that as well. Otherwise, this is good to go :)

File: CMakeLists.txt [hunk 14]
```diff
@@ -395,25 +453,28 @@ if(LIBUV_BUILD_TESTS)
            WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
 endif()
 
-if(UNIX)
+if (UNIX)
```
Comment (bnoordhuis) [inline CMakeLists.txt]:
```suggestion
if(UNIX)
```

File: CMakeLists.txt [hunk 15]
```diff
@@ -395,25 +453,28 @@ if(LIBUV_BUILD_TESTS)
...
   # Now for some gibbering horrors from beyond the stars...
-  foreach(x ${uv_libraries})
-    set(LIBS "${LIBS} -l${x}")
-  endforeach(x)
+  foreach(lib IN LISTS uv_libraries)
+    list(APPEND LIBS "-l${lib}")
+  endforeach()
+  string(REPLACE ";" " " LIBS "${LIBS}")
+  #Consider setting project version via project() call?
```
Comment (bnoordhuis) [inline CMakeLists.txt]:
```suggestion
  # Consider setting project version via project() call?
```

File: CMakeLists.txt [hunk 16]
```diff
@@ -395,25 +453,28 @@ if(LIBUV_BUILD_TESTS)
...
   file(STRINGS configure.ac configure_ac REGEX ^AC_INIT)
-  string(REGEX MATCH [0-9]+[.][0-9]+[.][0-9]+ PACKAGE_VERSION "${configure_ac}")
-  string(REGEX MATCH ^[0-9]+ UV_VERSION_MAJOR "${PACKAGE_VERSION}")
+  string(REGEX MATCH "([0-9]+)[.][0-9]+[.][0-9]+" PACKAGE_VERSION "${configure_ac}")
+  set(UV_VERSION_MAJOR "${CMAKE_MATCH_1}")
```
Comment (bnoordhuis) [inline CMakeLists.txt]: For my edification: why this change?

File: CMakeLists.txt [hunk 17]
```diff
@@ -395,25 +453,28 @@ if(LIBUV_BUILD_TESTS)
...
   set(includedir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR})
   set(libdir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR})
   set(prefix ${CMAKE_INSTALL_PREFIX})
-  configure_file(libuv.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc @ONLY)
+  configure_file(libuv.pc.in libuv.pc @ONLY)
 
   install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
   install(FILES LICENSE DESTINATION ${CMAKE_INSTALL_DOCDIR})
-  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc
+  install(FILES ${PROJECT_BINARY_DIR}/libuv.pc
```
Comment (bnoordhuis) [inline CMakeLists.txt]: Is PROJECT_BINARY_DIR correct when configure_file() is relative to CMAKE_CURRENT_BINARY_DIR?

Comment (slurps-mad-rips) [reply CMakeLists.txt]: In this case, yes. `${PROJECT_BINARY_DIR}` and `${CMAKE_CURRENT_BINARY_DIR}` are the same.

File: CMakeLists.txt (hunk 13 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: Looks like this can be in the next PR too so I can remind myself 😅

File: CMakeLists.txt (hunk 16 above)
Comment (slurps-mad-rips) [reply CMakeLists.txt]: This is here because the CMake regex engine is VERY poorly implemented. It's effectively taken from a public domain posted to a BBS in 1986 and hasn't changed since (it also can't be updated because then behavior would change 😓). Performing two regex calls here is also unnecessary, since after a `REGEX MATCH` call, CMake will set several `CMAKE_MATCH_<n>` variables. By putting a capture group at the front (where the major version is), we can do just the one call and then do a less expensive operation of setting the capture group.

It doesn't affect libuv itself, but anyone who uses `add_subdirectory` to include libuv in their build will get a small speed increase if their build is large enough. 

Though, the fastest thing would be to put the version inside the call to `project()` and update it as necessary.

File: CMakeLists.txt (hunk 17 above)
Comment (bnoordhuis) [reply CMakeLists.txt]: Okay, but why change it then? CMAKE_CURRENT_BINARY_DIR looks a little more self-evidently correct to me.

(Consulting the cmake docs didn't really give a satisfying answer.)

Comment (slurps-mad-rips) [reply CMakeLists.txt]: This might have been part of a `%s/CMAKE_CURRENT_BINARY/PROJECT_BINARY/g` command in vim. 

`CMAKE_CURRENT_BINARY_DIR` is set to whatever is the active CMakeLists.txt file.
`PROJECT_BINARY_DIR` is set to the current directory that called `project()` (This means that child directories can use `PROJECT_SOURCE_DIR` and friends and not affect parent scopes)

Typically you *rarely* need `CMAKE_CURRENT`, unless you take the more `autoconf` approach to CMakeLists.txt by putting one in each directory. If I was writing functions for general usage for a larger project (something on the scale of LLVM or GNU Radio, some monorepo project, or a generic CMake library), it would be something I would take into account, but for something as "small" in focus as libuv it's perfectly harmless.

Comment (bnoordhuis) [reply CMakeLists.txt]: Okay thanks, answer accepted. =)

**Comment by bnoordhuis** [comment]
https://ci.nodejs.org/job/libuv-test-commit/1732/

**Comment by slurps-mad-rips** [comment]
@bnoordhuis Looks like the ninja + cmake generator (Which is my default on windows, even with MSVC) didn't catch the `$<0>` bug that the visual studio generator found. I just pushed a fix.

There's other build issues from the non-cmake builds but I'm *pretty* sure that nothing I did caused it (famous last words, I know 😅)

**Comment by bnoordhuis** [comment]
New CI: https://ci.nodejs.org/job/libuv-test-commit/1733/

**Comment by richardlau** [comment]
> There's other build issues from the non-cmake builds but I'm _pretty_ sure that nothing I did caused it (famous last words, I know 😅)

The [aix-xlc](https://ci.nodejs.org/job/libuv-test-commit-aix/1768/nodes=aix72-xlc-ppc64/) failure is because the branch for this PR is 86 commits behind the current `v1.x` and so doesn't contain https://github.com/libuv/libuv/pull/2512.

**Comment by slurps-mad-rips** [comment]
@bnoordhuis I'll do a rebase onto latest before we get this merged in.

@richardlau thanks for the heads up on that!

**Comment by slurps-mad-rips** [comment]
@bnoordhuis rebased onto latest v1.x! 🙂

**Comment by bnoordhuis** [comment]
Thanks! CI: https://ci.nodejs.org/job/libuv-test-commit/1737/

**Comment by slurps-mad-rips** [comment]
@bnoordhuis looks like everything is green :)

Feel free to squash + merge or what have you :D

**Comment by saghul** [comment]
Squashed and landed in 018363a1. Thanks for the amazing work here @slurps-mad-rips! ✨ 🚀
